lib/simp/beaker_helpers.rb

```diff
@@ -693,40 +693,88 @@ def pluginsync_on( suts = hosts )
   #
   def latest_puppet_agent_version_for( puppet_version )
     return nil if puppet_version.nil?
+
     require 'rubygems/requirement'
     require 'rubygems/version'
     require 'yaml'
+
+    _puppet_version = puppet_version.strip
+
     @agent_version_table ||= YAML.load_file(
                                File.expand_path(
                                  '../../files/puppet-agent-versions.yaml',
                                  File.dirname(__FILE__)
                              )).fetch('version_mappings')
     _pair = @agent_version_table.find do |k,v|
-      Gem::Requirement.new(puppet_version).satisfied_by?(Gem::Version.new(k))
+      Gem::Requirement.new(_puppet_version).satisfied_by?(Gem::Version.new(k))
     end
     result = _pair ? _pair.last : nil
 
+    # If we didn't get a match, go look for published rubygems
+    unless result
+      puppet_gems = nil
+
+      Bundler.with_clean_env do
+        puppet_gems = %x(gem search -ra -e puppet).match(/\((.+)\)/)
+      end
+
+      if puppet_gems
+        puppet_gems = puppet_gems[1].split(/,?\s+/).select{|x| x =~ /^\d/}
+
+        # If we don't have a full version string, we need to massage it for the
+        # match.
+        if (_puppet_version.count('.') < 2) && !_puppet_version.include?(' ')
+          _puppet_version = "~> #{_puppet_version}"
+        end
+
+        result = puppet_gems.find do |ver|
+          Gem::Requirement.new(_puppet_version).satisfied_by?(Gem::Version.new(ver))
+        end
+      end
+    end
+
     return result
   end
 
 
   # Replacement for `install_puppet` in spec_helper_acceptance.rb
   def install_puppet
+    # The first match is internal Beaker and the second is legacy SIMP
+    puppet_agent_version = ENV['PUPPET_INSTALL_VERSION'] || ENV['PUPPET_VERSION']
+
+    # In case Beaker needs this down the line
+    ENV['PUPPET_INSTALL_VERSION'] = puppet_agent_version
```

Comment:
why is the ENV update done here and not after latest_puppet_agent_version_for() is called on line 773?